lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+
+  SmallVector<int64_t> srcShape;
+  auto dstShape = dst.getType().cast<ShapedType>().getShape();
+
+  if (src.getType().isa<ShapedType>()) {
+    srcShape =
+        SmallVector<int64_t>(src.getType().cast<ShapedType>().getShape());
+    assert(srcShape.size() == dstShape.size() &&
+           "rank of source and destination should match");
+
+    if (visitOperand(src, state, loc, builder).failed()) {
+      return failure();
+    }
+
+    for (size_t i = 0; i < dstShape.size(); i++) {
+      if (srcShape[i] == dstShape[i]) {
+        continue;
+      } else if (srcShape[i] < dstShape[i]) {
+        state.sizes[i] = builder.getIndexAttr(dstShape[i]);
+      } else {
+        llvm_unreachable("unexpected dimensions used in broadcast");
+      }
+    }
+    return success();
+  }
+
+  broadcastOp->emitRemark("PtrAnalysis: Unsupported broadcast source type");
+  return failure();
+}
+
+LogicalResult PtrAnalysis::visitOperandSplat(triton::SplatOp splatOp,
+                                             PtrSState &state,
+                                             const Location loc,
+                                             OpBuilder &builder) {
+  assert(state.isEmpty());
+
+  auto src = splatOp.getSrc();
+  auto dst = splatOp.getResult();
+  auto dstShape = dst.getType().cast<ShapedType>().getShape();
+
+  if (visitOperand(src, state, loc, builder).failed()) {
+    return failure();
+  }
+
+  if (src.getType().isa<IntegerType, IndexType, triton::PointerType>()) {
+    for (auto s : dstShape) {
+      state.offsets.push_back(builder.getIndexAttr(0));
+      state.sizes.push_back(builder.getIndexAttr(s));
+      state.strides.push_back(builder.getIndexAttr(0));
+      state.modulos.push_back(builder.getIndexAttr(0));
+    }
+  } else {
+    splatOp->emitRemark("PtrAnalysis: unsupported splat pattern");
+    return failure();
+  }
+
+  // If we splat a integer value, scalar should become the offset of the outer
+  // most dimension
+  if (state.scalar)
+    state.offsets[0] = state.scalar;
+
+  if (state.hasModulo() && state.getRank() > 2) {
+    splatOp->emitRemark("PtrAnalysis: unsupported scenario where splat result "
+                        "has modulo and rank > 2");
+    return failure();
+  }
+
+  return success();
+}
+
+LogicalResult PtrAnalysis::visitOperandAddptr(triton::AddPtrOp addptrOp,
+                                              PtrSState &state,
+                                              const Location loc,
+                                              OpBuilder &builder) {
+  assert(state.isEmpty());
+
+  PtrSState ptrState;
+  if (visitOperand(addptrOp.getPtr(), ptrState, addptrOp.getLoc(), builder)
+          .failed()) {
+    return failure();
+  }
+
+  PtrSState offsetState;
+  if (visitOperand(addptrOp.getOffset(), offsetState, addptrOp.getLoc(),
+                   builder)
+          .failed()) {
+    return failure();
+  }
+
+  assert(ptrState.source && "ptr field should provide source / base pointer");
```

Comment:
Similar question, I see this block gets removed too:

```cpp
  // Handle the special case when we are in a for loop, ptr is originally a
  // scalar pointer but replaced with a memref. In this case, ptrState will have
  // rank 1 and offsetState will have rank 0.
  // TODO:
  //  Passing a block argument pointer directly into a for loop not supported
  if (ptrState.getRank() == 1 && offsetState.getRank() == 0) {
    offsetState.sizes.push_back(rewriter.getIndexAttr(1));
    offsetState.offsets.push_back(offsetState.scalar);
    offsetState.strides.push_back(rewriter.getIndexAttr(0));
    offsetState.modulos.push_back(std::nullopt);
  }
```

Is this something we won't support

Reply:
This scenario is covered by `addptr_scalar_for_2d.mlir`. We still support it just by using slightly different approach. See above.